pytorch_lightning/metrics/functional/auroc.py

```diff
@@ -0,0 +1,179 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from distutils.version import LooseVersion
+from typing import Optional, Sequence, Tuple
+
+import torch
+
+from pytorch_lightning.metrics.classification.helpers import _input_format_classification
+from pytorch_lightning.metrics.functional.auc import auc
+from pytorch_lightning.metrics.functional.roc import roc
+from pytorch_lightning.utilities import LightningEnum
+
+
+class AverageMethods(LightningEnum):
+    """ Type of averages """
+    MACRO = 'macro'
+    WEIGHTED = 'weighted'
+    NONE = None
+
+
+def _auroc_update(preds: torch.Tensor, target: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor, str]:
+    # use _input_format_classification for validating the input and get the mode of data
+    _, _, mode = _input_format_classification(preds, target)
+
+    if mode == 'multi class multi dim':
+        n_classes = preds.shape[1]
+        preds = preds.transpose(0, 1).reshape(n_classes, -1).transpose(0, 1)
+        target = target.flatten()
+    if mode == 'multi-label' and preds.ndim > 2:
+        n_classes = preds.shape[1]
+        preds = preds.transpose(0, 1).reshape(n_classes, -1).transpose(0, 1)
+        target = target.transpose(0, 1).reshape(n_classes, -1).transpose(0, 1)
+
+    return preds, target, mode
+
+
+def _auroc_compute(
+        preds: torch.Tensor,
+        target: torch.Tensor,
+        mode: str,
+        num_classes: Optional[int] = None,
+        pos_label: Optional[int] = None,
+        average: Optional[str] = 'macro',
+        max_fpr: Optional[float] = None,
+        sample_weights: Optional[Sequence] = None,
+) -> torch.Tensor:
+    # binary mode override num_classes
+    if mode == 'binary':
+        num_classes = 1
+
+    # check max_fpr parameter
+    if max_fpr is not None:
+        if (not isinstance(max_fpr, float) and 0 < max_fpr <= 1):
+            raise ValueError(f"`max_fpr` should be a float in range (0, 1], got: {max_fpr}")
+
+        if LooseVersion(torch.__version__) < LooseVersion('1.6.0'):
+            raise RuntimeError("`max_fpr` argument requires `torch.bucketize` which"
+                               " is not available below PyTorch version 1.6")
+
+        # max_fpr parameter is only support for binary
+        if mode != 'binary':
+            raise ValueError(f"Partial AUC computation not available in "
+                             f"multilabel/multiclass setting, 'max_fpr' must be"
+                             f" set to `None`, received `{max_fpr}`.")
+
+    # calculate fpr, tpr
+    if mode == 'multi-label':
```

Comment:
lest do it in another PR, but also move this to Enum :]

Reply:
Agree :]
I think multiple places in metrics could be converted into Enum

Reply:
cool, co lets create an issue for and mark is good-first-issue :]

Reply:
I do that :]